File: PythonScheduler/scheduler.py
import random
from employee import Employee

DAYS = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]
SHIFTS = ["Morning", "Afternoon", "Evening"]

class Scheduler:
    def __init__(self):
        self.employees = []
        self.schedule = {day: {shift: [] for shift in SHIFTS} for day in DAYS}

    def add_employee(self, employee):
        self.employees.append(employee)

    def collect_preferences(self):
        n = int(input("Enter number of employees: "))
        for _ in range(n):
            name = input("Enter employee name: ")
            emp = Employee(name)
            for day in DAYS:
                pref = input(f"Preferred shift for {day} (Morning/Afternoon/Evening): ").capitalize()
                emp.preferred_shifts[day] = pref if pref in SHIFTS else "Morning"
            self.add_employee(emp)

    def assign_shifts(self):
        for day in DAYS:
            for shift in SHIFTS:
                available = [e for e in self.employees if e.can_work(day) and e.preferred_shifts.get(day) == shift]
                assigned = available[:2]  # take up to 2 preferred
                for emp in assigned:
                    emp.assign_shift(day, shift)
                    self.schedule[day][shift].append(emp.name)

                # If fewer than 2, randomly assign others
                while len(self.schedule[day][shift]) < 2:
                    others = [e for e in self.employees if e.can_work(day) and e.name not in self.schedule[day][shift]]
                    if not others:
                        break
                    chosen = random.choice(others)
                    chosen.assign_shift(day, shift)
                    self.schedule[day][shift].append(chosen.name)

    def print_schedule(self):
        print("\nFinal Weekly Schedule:")
        for day in DAYS:
            print(f"\n{day}")
            for shift in SHIFTS:
                assigned = ", ".join(self.schedule[day][shift])
                print(f"  {shift}: {assigned if assigned else 'No one assigned'}")

if __name__ == "__main__":
    s = Scheduler()
    s.collect_preferences()
    s.assign_shifts()
    s.print_schedule()

File: CSharpScheduler/Scheduler.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ShiftScheduler
{
    public class Scheduler
    {
        private static readonly string[] DAYS = { "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday" };
        private static readonly string[] SHIFTS = { "Morning", "Afternoon", "Evening" };

        private List<Employee> Employees = new();
        private Dictionary<string, Dictionary<string, List<string>>> Schedule = new();

        public Scheduler()
        {
            foreach (var day in DAYS)
            {
                Schedule[day] = new Dictionary<string, List<string>>();
                foreach (var shift in SHIFTS)
                {
                    Schedule[day][shift] = new List<string>();
                }
            }
        }

        public void CollectPreferences()
        {
            Console.Write("Enter number of employees: ");
            int n = int.Parse(Console.ReadLine());

            for (int i = 0; i < n; i++)
            {
                Console.Write("Enter employee name: ");
                var name = Console.ReadLine();
                var emp = new Employee(name);

                foreach (var day in DAYS)
                {
                    Console.Write($"Preferred shift for {day} (Morning/Afternoon/Evening): ");
                    var pref = Console.ReadLine();
                    emp.PreferredShifts[day] = SHIFTS.Contains(pref, StringComparer.OrdinalIgnoreCase) ? pref : "Morning";
                }

                Employees.Add(emp);
            }
        }

        public void AssignShifts()
        {
            var rand = new Random();

            foreach (var day in DAYS)
            {
                foreach (var shift in SHIFTS)
                {
                    var preferred = Employees
                        .Where(e => e.CanWork(day) && string.Equals(e.PreferredShifts[day], shift, StringComparison.OrdinalIgnoreCase))
                        .Take(2)
                        .ToList();

                    foreach (var emp in preferred)
                    {
                        emp.AssignShift(day, shift);
                        Schedule[day][shift].Add(emp.Name);
                    }

                    while (Schedule[day][shift].Count < 2)
                    {
                        var available = Employees
                            .Where(e => e.CanWork(day) && !Schedule[day][shift].Contains(e.Name))
                            .ToList();

                        if (available.Count == 0) break;

                        var fallback = available[rand.Next(available.Count)];
                        fallback.AssignShift(day, shift);
                        Schedule[day][shift].Add(fallback.Name);
                    }
                }
            }
        }

        public void PrintSchedule()
        {
            Console.WriteLine("\nFinal Weekly Schedule:");
            foreach (var day in DAYS)
            {
                Console.WriteLine($"\n{day}:");
                foreach (var shift in SHIFTS)
                {
                    var names = Schedule[day][shift];
                    Console.WriteLine($"  {shift}: {(names.Count > 0 ? string.Join(", ", names) : "No one assigned")}");
                }
            }
        }
    }
}
